CMakeLists.txt

```diff
@@ -6,138 +6,555 @@
 
-# LLVM_LIT_EXTERNAL store in cache so it could be used by AddLLVM.cmake
-set(LLVM_EXTERNAL_LIT ${LLVM_TOOLS_BINARY_DIR}/llvm-lit CACHE STRING "Command used to spawn lit")
+  option(LLVM_FORCE_USE_OLD_HOST_TOOLCHAIN
+    "Set to ON to force using an old, unsupported host toolchain." OFF)
+  option(CLANG_ENABLE_BOOTSTRAP "Generate the clang bootstrap target" OFF)
+  option(LLVM_ENABLE_LIBXML2 "Use libxml2 if available." ON)
+
+  include(AddLLVM)
+  include(TableGen)
+  include(HandleLLVMOptions)
+  include(VersionFromVCS)
+# FIXME: uncomment once FIR code is upstreamed. 
+#  include(AddMLIR)
+
+  set(PACKAGE_VERSION "${LLVM_PACKAGE_VERSION}")
+
+  if (NOT DEFINED LLVM_INCLUDE_TESTS) 
+    set(LLVM_INCLUDE_TESTS ON)
+  endif()
+
+  include_directories("${LLVM_BINARY_DIR}/include" "${LLVM_MAIN_INCLUDE_DIR}")
+  link_directories("${LLVM_LIBRARY_DIR}")
+  set(MLIR_BINARY_DIR ${LLVM_BINARY_DIR})
+
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+  set(CMAKE_LIBRARY_OUTPUT_DIRECTORY 
+    ${CMAKE_BINARY_DIR}/lib${LLVM_LIBDIR_SUFFIX})
+  set( CMAKE_ARCHIVE_OUTPUT_DIRECTORY 
+    ${CMAKE_BINARY_DIR}/lib${LLVM_LIBDIR_SUFFIX})
+  
+  if (LLVM_INCLUDE_TESTS)
+    include(FindPythonInterp)
+    if (NOT PYTHONINTERP_FOUND)
+      message(FATAL_ERROR
+        "Unable to find Python interpreter, required for builds and testing. \
+        Please install Python or specify the PYTHON_EXECUTABLE CMake variable.")
+    endif()
+
+    if (${PYTHON_VERSION_STRING} VERSION_LESS 2.7)
+      message(FATAL_ERROR "Python 2.7 or newer is required.")
+    endif()
+  
+    # Check prebuilt llvm/utils.
+    if (EXISTS ${LLVM_TOOLS_BINARY_DIR}/FileCheck${CMAKE_EXECUTABLE_SUFFIX}
+        AND EXISTS ${LLVM_TOOLS_BINARY_DIR}/count${CMAKE_EXECUTABLE_SUFFIX}
+        AND EXISTS ${LLVM_TOOLS_BINARY_DIR}/not${CMAKE_EXECUTABLE_SUFFIX})
+      set(LLVM_UTILS_PROVIDED ON)
+    endif()
+
+    if (EXISTS ${LLVM_MAIN_SRC_DIR}/utils/lit/lit.py)
+      # Note: the path is not really used, except for checking if lit was found
+      set(LLVM_LIT ${LLVM_MAIN_SRC_DIR}/utils/lit/lit.py)
+      if(EXISTS ${LLVM_MAIN_SRC_DIR}/utils/llvm-lit)
+        add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/llvm-lit utils/llvm-lit)
+      endif()
+
+    if(NOT LLVM_UTILS_PROVIDED)
+      add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/FileCheck utils/FileCheck)
+      add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/count utils/count)
+      add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/not utils/not)
+      set(LLVM_UTILS_PROVIDED ON)
+      set(FLANG_TEST_DEPS FileCheck count not)
+    endif()
+
+    set(UNITTEST_DIR ${LLVM_MAIN_SRC_DIR}/utils/unittest)
+    
+    if (EXISTS ${UNITTEST_DIR}/googletest/include/gtest/gtest.h
+        AND NOT EXISTS 
+          ${LLVM_LIBRARY_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}gtest${CMAKE_STATIC_LIBRARY_SUFFIX}
+        AND EXISTS ${UNITTEST_DIR}/CMakeLists.txt)
+          add_subdirectory(${UNITTEST_DIR} utils/unittest)
+      endif()
```

Comment:
Is there any plan to use this immediately? If not, let's leave it out until it is needed.

Reply:
The Arm folks had expressed some interest in getting this working as part of the process.  Not sure if it suits what they were looking for but left it here for them to provide feedback and/or build upon.

Reply:
I agree with @tskeith here. We can leave this out of the first cut to keep it clean and bring it back in if/when we add gtest-style tests similar to the unit tests in clang.